LAN games discovery #8906

rob-v opened the pull request:
Used library https://github.com/rix0rrr/beacon

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -25,6 +25,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 {
 	public class LobbyLogic : ChromeLogic
 	{
+		public const string OpenRALANGame = "OpenRALANGame";
```
The logic here really needs to be in the `MasterServerPinger` server trait.  Putting it in the `LobbyLogic` means that games will only be advertised after the first player has joined.  This will also fix the issue where games in progress aren't advertised.

rob-v replied:
It is advertised after host creates game so also before the first player joins and also after each change in map or players/bots. Re 'Started' games I didn't consider it a real issue as we are mainly interested in Waiting games for LAN. Advantage of LobbyLogic is, that it is all in one place with all information (start, change of map, change of players,.). I will move it to MasterServerPinger. Do you want to keep advertising also Started games?

pchote replied:
`LobbyLogic` runs client side for every player who joins a server, not on the server itself.  This distinction is important for dedicated servers.

Yes, please do advertise started games.

pchote commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -37,6 +37,9 @@
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="BeaconLib">
+      <HintPath>..\thirdparty\download\BeaconLib.dll</HintPath>
```
This will need to be properly integrated with our nuget dependency fetching.

rob-v replied:
How/who can do it?

pchote replied:
You'll need to update the scripts in /thirdparty for this.  You will also need to update the makefile so that this will compile on non-windows systems.

rob-v commented:
Updated - LAN game advertisement moved to MasterServerPinger, works also for Started games.

atlimit8 commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -826,4 +821,4 @@
     -->
   </Target>
-</Project>
+</Project>
```
Could you leave the new line here?

rob-v replied:
It was done by VS 2015. I will revert it, but if it is done by VS each time, it will be always manual action.

atlimit8 commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -802,22 +805,14 @@
-      DestinationFolder="$(SolutionDir)mods/common/"
-      Condition="'$(OS)' == 'Unix'"/>
+    <Copy SourceFiles="$(TargetDir)$(TargetName).pdb" DestinationFolder="$(SolutionDir)mods/common/" Condition="'$(OS)' == 'Windows_NT'" />
+    <Copy SourceFiles="$(TargetPath).mdb" DestinationFolder="$(SolutionDir)mods/common/" Condition="'$(OS)' == 'Unix'" />
```
I don't think that these lines should be changed. Their current form is more readable.

rob-v replied:
It was done by VS 2015. I will revert it, but if it is done by VS each time, it will be always manual action.

atlimit8 commented:
This PR fails continuous integration tests. I don't think it will compile for others. Try making a clean clone of your branch in a different location and building that.

See https://travis-ci.org/OpenRA/OpenRA/builds/222360145 and https://ci.appveyor.com/project/OpenRA/openra/build/1.0.15108

rob-v commented:
@atlimit8 Can it be related to new .dll? Phrohdoh created issue for that library to create nuget package. Then I need to update some .sh/makefile files.

rob-v commented:
Updated - fixed .csproj formatting.

pchote commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -23,34 +24,48 @@ public class MasterServerPinger : ServerTrait, ITick, INotifySyncLobbyInfo, ISta
 		// 3 minutes. Server has a 5 minute TTL for games, so give ourselves a bit of leeway.
 		const int MasterPingInterval = 60 * 3;
+		public const string OpenRALANGame = "OpenRALANGame";
+		readonly int currentProcessId = System.Diagnostics.Process.GetCurrentProcess().Id;
```
It would probably be worth moving this to a static property in `Platform`.

pchote commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -23,34 +24,48 @@ public class MasterServerPinger : ServerTrait, ITick, INotifySyncLobbyInfo, ISta
 		Queue<string> masterServerMessages = new Queue<string>();
 
-		public void PingMasterServer(S server)
+		public void PingMasterServer(S server, bool createLANGameBeacon)
```
This method now feels a bit too spaghettiish.   I suggest pulling the `numPlayers` etc out into another method that then calls this and `CreateLANGameBeacon` individually.

pchote commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -115,5 +138,33 @@ public void PingMasterServer(S server)
+	Protected: {11}".F(currentProcessId, settings.Name, settings.ListenPort, (int)server.State, numPlayers, numSlots, numBots, numSpectators,
+				server.Map.Uid, mod.Id, mod.Metadata.Version, !string.IsNullOrEmpty(settings.Password));
+
+			if (lanGameYaml != lastLanGameBeacon)
```
It would be a lot cleaner to check for state changes at a higher level (or not - is there a downside to recreating beacons periodically?) and then have that apply to both LAN and master server updates.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs:
```diff
@@ -106,6 +111,11 @@ public MultiplayerLogic(Widget widget, ModData modData, Action onStart, Action o
 			Game.LoadWidget(null, "GLOBALCHAT_PANEL", widget.Get("GLOBALCHAT_ROOT"), new WidgetArgs());
 
+			lanGameLocations = new List<BeaconLocation>();
+			lanGameProbe = new Probe(Server.MasterServerPinger.OpenRALANGame);
```
Referencing the constant from the server code is quite ugly.  Can we please hardcode "OpenRALANGame" in both places instead?

rob-v replied:
Ok, though to not make it ugly but still maintainable, I would create e.g. Constants class where we could put any such constants with no clear place.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs:
```diff
@@ -106,6 +111,11 @@ public MultiplayerLogic(Widget widget, ModData modData, Action onStart, Action o
 
+			lanGameLocations = new List<BeaconLocation>();
+			lanGameProbe = new Probe(Server.MasterServerPinger.OpenRALANGame);
+			lanGameProbe.BeaconsUpdated += locations => lanGameLocations = locations;
```
You will need to refresh the server list, too.

I suggest changing this and `RefreshServerList` up to update their own `List<GameServer>`s, and then change `RefreshServerListInner` to work with those lists instead of a passed in argument.

rob-v replied:
> You will need to refresh the server list, too.
This isn't clear to me. You refresh server list after button click in onComplete. At the same time it retrieves server games it appends local games so the window is refreshed for both server and local games.

pchote reviewed (changes requested):
The overall approach looks good, just a couple of suggestions/requests to make the logic simpler and more maintainable.

Commit:
LAN games discovery

rob-v commented:
Updated.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs:
```diff
@@ -322,7 +329,13 @@ void RefreshServerList()
-				Game.RunAfterTick(() => RefreshServerListInner(games));
+				var lanGames = lanGameLocations.Select(bl => {
+						var gs = new GameServer(MiniYaml.FromString(bl.Data)[0].Value);
+						gs.Address = bl.Address.ToString().Split(':')[0] + ":" + gs.Address.Split(':')[1];
```
Making the address non-readonly and then manipulating it here isn't very nice.  Please modify the yaml given to the GameServer instead:
```csharp
bl =>
{
	var game = MiniYaml.FromString(bl.Data)[0].Value;

	// Rewrite the server address with the correct IP
	var addressNode = game.Nodes.FirstOrDefault(n => n.Key == "Address");
	if (addressNode != null)
		addressNode.Value.Value = bl.Address.ToString().Split(':')[0] + ":" + addressNode.Value.Value.Split(':')[1];

	return new GameServer(game);
}
```

Also note a style nit: opening braces for a block should be on their own line.

pchote commented on OpenRA.Game/Network/GameServer.cs:
```diff
@@ -15,7 +15,6 @@ public class GameServer
 	{
 		public readonly int Id = 0;
 		public readonly string Name = null;
-		public readonly string Address = null;
```
You can then drop the changes to this file.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs:
```diff
@@ -322,7 +329,13 @@ void RefreshServerList()
+						var gs = new GameServer(MiniYaml.FromString(bl.Data)[0].Value);
+						gs.Address = bl.Address.ToString().Split(':')[0] + ":" + gs.Address.Split(':')[1];
+						return gs;
+					}).Where(g => g.Id != Platform.CurrentProcessId).GroupBy(gs => gs.Address).Select(g => g.Last()).ToList();
```
How hard would it be to do this filtering off the beacons rather than the GameServers?  That would more cleanly separate the "find games on the network" vs "turn those games into server entries" steps.

pchote commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -116,5 +146,26 @@ public void PingMasterServer(S server)
+		{
+			var settings = server.Settings;
+			var lanGameYaml =
+@"Game:
```
Can you please add a `// TODO: Serialize and send client names` here?
This is a stub of a feature that has only been partly implemented, but it would be good to acknowledge that this bit will need to be updated in the future.  There is no point in actually implementing it now, I expect the details of the serialization will need to change.

pchote commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -23,35 +24,52 @@ public class MasterServerPinger : ServerTrait, ITick, INotifySyncLobbyInfo, ISta
+
 		public void Tick(S server)
 		{
+			if (firstTick)
```
You can implement `INotifyServerStart` and then put this inside `INotifyServerStart.ServerStarted(Server s)`.

pchote commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -23,35 +24,52 @@ public class MasterServerPinger : ServerTrait, ITick, INotifySyncLobbyInfo, ISta
+			PublishGames(server);
+		}
 
+		void PublishGames(S server)
```
This is only publishing one game, so `PublishGame`?

pchote commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -60,9 +78,21 @@ public void PingMasterServer(S server)
+			UpdateLANGameBeacon(server, numPlayers, numSlots, numBots, numSpectators, mod, passwordProtected);
+		}
+
+		void PingMasterServer(S server, int numPlayers, int numSlots, int numBots, int numSpectators, Manifest mod, bool passwordProtected, string[] clients)
```
Could you rename this to `UpdateMasterServer` for consistency?

pchote reviewed (changes requested):
This works really well, good job!  I have a few more requests below, but then I think this should be good to merge.  We should also get :+1:'s from people on Windows and Linux before we merge.

I notice a few "System.Net.Sockets.SocketOptionName 0x17 is not supported at IP level" being written to stdout by, i assume, beaconlib, but that doesn't appear to stop it from working.

Finally, it would be nice for polish if we could change the "Country" column in the server browser to "Location", and then have local games show as "Local Network" instead of "Unknown Loca...".  
The easiest way will be to check if the server IP matches the reserved internal IP ranges, but in any case this is out of scope for this PR and should be left to a followup.

rob-v commented:
Updated as requested but with some additional changes to be reviewed.
I replaced ProcessId by Guid to ensure it is unique. Due to this change to replace then Id by -1, I used this to identify 'Local Network' games. We could also add GameServer.IsLANGame flag rather than checking the IP range.

pchote commented:
Ah, well actualy the reason why I suggested to check the IP range and also do it in a separate PR was so that it could also fix the issue where the player location in the lobby tooltips show "Unknown Location".

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -36,6 +36,8 @@ public static class Game
 		public const int Timestep = 40;
 		public const int TimestepJankThreshold = 250; // Don't catch up for delays larger than 250ms
 
+		public static readonly Guid GameId = Guid.NewGuid();
```
Again, the server code should avoid referring to anything in the Game class because dedicated servers run without a game client.  This doesn't change anything in practice, but is a code smell and potential maintenance issue.

Moving this to a `Platform.SessionGUID` field should be sufficient.

pchote reviewed (approved):
Looks good otherwise :+1:

rob-v commented:
Updated, thanks.

reaperrr commented:
Needs a rebase.

rob-v commented:
Rebased.

pchote commented:
Needs another rebase, sorry.

@OpenRA/engine-hackers: this has now been stalled waiting for a second :+1: for nearly three weeks.  This really isn't hard to review.

reaperrr commented:
Seems this was already rebased. Looks good to me and works fine on Windows, but before we merge this, is the AppVeyor fail a one-time thing or does that need som kind of fix @pchote ?

pchote commented:
I think that is another case of #12504 (which was only fixed for the bash scripts) rather than something wrong with this PR.  It will probably need a fix, but not here.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/50b7d3994f9024d17d867c3060f9dd8e74fc657c...46201633a16ed1ff51ea763357f21f102d7d88f3
